### ISSUE TITLE
smartparens modern Fortran support

### ISSUE THREAD

**Issue by zbeekman** [Fuco1/smartparens#304]
Hi,
I'm new to smartparens, and in general, not a great lisp/Elisp coder. I work a lot with modern Fortran (F03/08) and would love to implement smartparens support in a smartparens-fortran.el file. Taking a look at the Ruby file, (Fortran uses block constructs somewhat similar to ruby or c) the chief difficulty seems to be in implementing the helper functions needed. That is, I think I understand the usage of sp-local-pair pretty well, but I'm not confident in my ability to implement the other elisp functions needed for the logic. Some example Fortran blocks are:

``` fortran
! The following are all block constructs
module my_module
end module

program my_program
end program

subroutine my_sub(arg1,arg2)
end subroutine

function my_fun(arg1,arg2) result(res)
end function

! Block creates a new scoping unit within executable portion and allows declaration of
! new local variables
block
end block

interface my_generic_interface
end interface

 ! There are also attributes possible here, and it might get tricky.
type my_type(parameter1,parameter2)
end type

! or

type ,extends(person) :: employee
end type

associate (x => user_type%x_component)
end associate

do i=1,10,2 !strided do loop
   ! Do something
end do

! The following may be block constructs OR statements so we might have to 
! do something clever with the :when argument to sp-local-pair

! Conflicting prefix with do loop, but maybe some magic possible with extra defuns?
! After all the closing elements are the same
do concurrent (i=1:10:2,j=0:jmx-1)
   ! Loops can be executed in any order, and potentially in parallel.
end do

forall (i=1:imax,j=1:jmax,j<=i)
    A_lower(i,j) = A(i,j)
end forall

where ( A /= 0 )
    A = 1/A !Masked array assignment
    B = A ! Won't get zeros/infs.
end where
! vs.
where ( A /= 0 ) A = 1/A

if ( a > b ) then
! Execute if true
end if
! vs
if ( a > b ) print*, 'A is greater than B'


! The rest must be block constructs only, not statements
! Potential issues between these next two b/c of common prefix
! Maybe some extra defuns could handle this since the closing element is the same
select case (var)
end select 

select type (var) 
end select
```

There may be more fortran specific pairs besides those globally available in smart parens, but I can't think of them at the moment. (Well I know of one, but it's not too common and does not satisfy the no-common-prefixes rule: array constructors `(/` `/)` but as of Fortran 2003 one can use square braces instead.)

I don't have time now, but I think I'm going to try to slog through [smartparens-ruby.el](../blob/master/smartparens-ruby.el) and really understand what all those declared functions are doing, and then use that as the basis for smartparens-fortran.el. Any help, or pointers would be very much appreciated. There are a number of functions declared in f90-mode which might be useful but many of them have the unfortunate side effect of adding the name to the end of named block constructs. (This is optional according to the standard, and Rousson et al make good arguments about why adding optional names to the closing portion of the blocks can be a bad idea; for example, if you change the name of a function you need to remember to update its name in the `end function fun_name` portion as well.)

**Comment by Fuco1** [comment]
Hi @zbeekman I'm doing a bit of a cleanup.  Have you hacked on something that would be interesting to get included in the project?  If not that's fine.

I'm going to close this issue now and if you have some possible contributions you can open a new issue or a pull request and we can work from there.  Thanks!
